### PR TITLE
Homework 29 completed

### PR DESCRIPTION
**Pull request by Borisko130**

### CODE REVIEW

File: src/main/java/com/internet/shop/model/User.java [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 
+import java.util.List;
 import java.util.Objects;
 
 public class User {
     private Long id;
     private String name;
     private String login;
     private String password;
+    private List<Role> roles;
```
Comment (gradiviro) [inline src/main/java/com/internet/shop/model/User.java:11]: В UML рекомендовали использовать Set
```suggestion
    private Set<Role> roles;
```

File: src/main/java/com/internet/shop/model/User.java [hunk 2]
```diff
@@ -46,6 +48,14 @@ public void setPassword(String password) {
         this.password = password;
     }
 
+    public List<Role> getRoles() {
```
Comment (gradiviro) [inline src/main/java/com/internet/shop/model/User.java:51]: same here

File: src/main/java/com/internet/shop/web/filter/AuthorizationFilter.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    private Map<String, List<Role.RoleName>> protectedUrls = new HashMap<>();
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+        protectedUrls.put("/users", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/products/add", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/products/manage", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/orders/manage", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/shopping-cart/products", List.of(Role.RoleName.USER));
+        protectedUrls.put("/orders/all", List.of(Role.RoleName.USER));
```
Comment (gradiviro) [inline src/main/java/com/internet/shop/web/filter/AuthorizationFilter.java:33]: В чек листе рекомендовали добавить все url из web.xml например удаление/добавление товара

**Review by gradiviro — APPROVED** [review]

File: src/main/java/com/internet/shop/web/filter/AuthorizationFilter.java (hunk 3 above)
Comment (dmytro-hryhoruk) [reply src/main/java/com/internet/shop/web/filter/AuthorizationFilter.java:33]: agree

File: src/main/java/com/internet/shop/model/User.java (hunk 1 above)
Comment (dmytro-hryhoruk) [reply src/main/java/com/internet/shop/model/User.java:11]: agree

File: src/main/java/com/internet/shop/model/User.java (hunk 2 above)
Comment (dmytro-hryhoruk) [reply src/main/java/com/internet/shop/model/User.java:51]: agree

File: src/main/java/com/internet/shop/controller/user/AddUserController.java [hunk 4]
```diff
@@ -39,6 +41,12 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp)
         } else if (pass.equals(passRep)) {
             User user = new User(name, login, pass);
             userService.create(user);
+            // If this user has id 1, then it will be ADMIN
+            if (user.getId() == 1L) {
+                user.setRoles(Set.of(Role.of("ADMIN")));
+            } else {
+                user.setRoles(Set.of(Role.of("USER")));
+            }
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/controller/user/AddUserController.java:49]: it works if you are using arraylists for storing users. But in situation with the real database you'll end up with user without roles in database(cause userService.create(user); invoked before setting roles to user)

**Review by dmytro-hryhoruk — APPROVED** [review]

File: src/main/java/com/internet/shop/controller/user/AddUserController.java (hunk 4 above)
Comment (dmytro-hryhoruk) [reply src/main/java/com/internet/shop/controller/user/AddUserController.java:49]: let's change it

File: src/main/java/com/internet/shop/controller/user/AddUserController.java [hunk 5]
```diff
@@ -38,6 +40,12 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp)
             req.getRequestDispatcher("/WEB-INF/views/users/addUser.jsp").forward(req, resp);
         } else if (pass.equals(passRep)) {
             User user = new User(name, login, pass);
+            // If this is first user, then it will be ADMIN
+            if (userService.getAll().size() == 0) {
+                user.setRoles(Set.of(Role.of("ADMIN")));
+            } else {
+                user.setRoles(Set.of(Role.of("USER")));
+            }
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/controller/user/AddUserController.java:48]: okey, that's gonna work but imagine that each time while creating a new user you will download the whole database(all users). That may be a really expensive operation. You may create admin in Inject Controller and just use him but here let's set User role by default

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**